ext/reflection/php_reflection.c

```diff
@@ -305,6 +313,8 @@ static void reflection_free_objects_storage(zend_object *object) /* {{{ */
 		case REF_TYPE_PARAMETER:
 			reference = (parameter_reference*)intern->ptr;
 			_free_function(reference->fptr);
+			/* fallthrough */
+		case REF_TYPE_TYPE:
```

Comment:
The problem with the return type was not on your side. I've just landed a change that should fix this and a similar problem for variadic parameters.

The fest for copy/free I had in mind is something like this:

```php
$closure = function(Test $x) {};
$rm = new ReflectionMethod($closure, '__invoke');
$rp = $rm->getParameters()[0];
$rt = $rp->getType();
unset($rm, $rp);
var_dump((string) $rt);
```

And then see if this causes memory errors in valgrind.